### DISCUSSION TITLE
low-level Network ideas

### DISCUSSION THREAD

**Discussion by GFW-knocker** [GFW-knocker/gfw_resist_tls_proxy#34] [Announcements]
dear friends
thanks for your contribution
please share your ideas here
specially [low-level] network related ideas

**Comment by GFW-knocker** [comment]
عزیرانی که در قسمت شبکه توانایی دارند

ایده و ابزار کاربردی  تست شده هر چی دارین اینجا بزارین
با نقض پروتکل میشه از فایروال عبور کرد
دستمون خیلی بازه
مثلا یهو بسته tcp رو بدون syn بفرستیم به سرور خودمون میرسه؟ 
اگه رد شد سرعتش چطوره؟
یا ایرانسل چطور سرعت رو کاهش میده به چی حساسه؟

**Reply by free-the-internet** [reply]
با ایرانسل، client hello با پارامترهای fragmentation شما به سرور میرسه؟ من یادم میاد در مورد روسیه، اگه مثلا double packet داشتیمتوی بافر فعلی، کلا drop میشد بافر. از این حهت میگم که شاید تایم اوت میشه بافر و کلا ریست میزنه به جای ارسال.

**Reply by GFW-knocker** [reply]
بله میرسه server hello  هم دریافت میشه حتی سرچ گوگل انجام میشه ولی خیلی خیلی کند
اینقدر کنده که تقریبا با قطع فرقی نداره
و رو همون ایپی سایت plos.org سریع باز میشه

**Reply by free-the-internet** [reply]
یا خود رنج آی پی رو کلا throttle میکنن که امکان نداره وقتی میگی plos.org به خوبی باز میشه.
یا اینکه وقتی بافر فایروال رو overflow میکنی کلا session کانکشن رو flag میکنن (البته من نمیدونم چقدر میتونن اینکارو بکنن. در حد حدس و گمان هست)
دوستان دیگه در داخل ایران تست های بیشتری انجام بدن، میشه به نظرم پیدا کرد که چکار دارن میکنن.

**Reply by free-the-internet** [reply]
یه علت دیگه هم ممکنه این باشه که پکت هایی که Client Hello شناخته میشن، اگه SNI شون به هر دلیلی تشخیص داده نشد رو دراپ نکنن، ولی کانکشن رو توی QoS رده های بسیار پایین بزارن. شاید بشه یک فرگمنت به آخر اضافه کرد حاوی فیلد SNI فیک (که جزوی از پکت اصلی نباشه و فقط جهت نشون دادن به فایروال باشه)

**Reply by free-the-internet** [reply]
توی لینوکس وقتی delay بین چند تا send کم باشه، هر چند شما fragmentation انجام داده باشید، ممکنه در کرنل باز هم reassembly رخ بده و بعد ارسال بشه. برای این موضوع شما باید از آپشن TCP_NODELAY در ساکتی که میسازی استفاده کنید. اطلاعات بیشتر: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/7/html/tuning_guide/tcp_nodelay_and_small_buffer_writes

@GFW-knocker

**Reply by GFW-knocker** [reply]
@free-the-internet 
special thanks to you
i add TCP_NODELAY and it make loading speed better , i guess

**Reply by GFW-knocker** [reply]
> یه علت دیگه هم ممکنه این باشه که پکت هایی که Client Hello شناخته میشن، اگه SNI شون به هر دلیلی تشخیص داده نشد رو دراپ نکنن، ولی کانکشن رو توی QoS رده های بسیار پایین بزارن. شاید بشه یک فرگمنت به آخر اضافه کرد حاوی فیلد SNI فیک (که جزوی از پکت اصلی نباشه و فقط جهت نشون دادن به فایروال باشه)

حدس منم QoS است
ولی فرض کنید رشته plos.org رو بخوایم بفرستیم که فریب بخوره
این رشته رو چطوری بفرستم که پروتکل نقض نشه
اونطرف سرور کلادفلر قرار داره و ما باید طوری بفرستیم که پروتکل tls و یا tcp نقض نشه

**Reply by free-the-internet** [reply]
همونطور که گفتم ببینید میشه یک فیلد SNI دیگه حاوی دامنه ی مجاز به آخر extension های پکت Client Hello اضافه کنید؟ طوری که اون فیلد SNI قابل تشخیص باشه قبل assembly (احتمالش کمه جواب بده این موضوع ولی میشه تست کرد)
یا اینکه، از Client Hello اصلی کپی بگیرید، اولی رو با Fragmentation بفرستید. دومی رو فقط فیلد SNI رو دستکاری کنید و بفرستید. احتمالا مشکل checksum پیدا میکنید که باید درست کنید. ولی من جای شما باشم قبل اینکه checksum رو درست کنم همینو میفرستم بدون fragmentation که فقط فایروال SNI رو ببینه و بعد بخاطر checksum اشتباه احتمالا دراپ کنه. اگه کار نکرد checksum رو هم درست کنید و تست کنید به امید اینکه کار کنه.

**Reply by free-the-internet** [reply]
البته این کار ممکنه کمی سخت باشه ولی بهرحال این مشکلات نیازمند کار در لایه های پایینی دارند که حل بشوند.

**Comment by free-the-internet** [comment]
تست کردین که آیا با این روش مستقیما میشه بعضی از سایت های فیلتر رو باز کرد یا نه؟
مثلا از فیلتر DNS برای یوتیوب که گذشتین (با هر روشی)، ببینی با استفاده از این اسکریپت باز میشه یا نه؟

**Reply by GFW-knocker** [reply]
مرسی. الان چک میکنم.

**Reply by free-the-internet** [reply]
درود بر شما
لطفا در readme صفحه ی اول این repository اسم یوزر منو به @free-the-internet  تغییر بدید و آن را به پرفایل من در گیت هاب لینک کنید که افراد بتوانند فالو کنند. اینطوری شاید community بهتری پیدا میکنم که بتوانم پروژه های مبارزه با فیلترشان رو ببینم و مشارکت کنم.
سپاس از زحمات شما.

**Comment by electron-v2ray** [comment]
تازه دیدم ولی روی ایرانسل کار نکرد

[SHOW](https://telegra.ph/%D8%B1%D8%B4%D8%AA%D9%87-%D8%AA%D9%88%DB%8C%DB%8C%D8%AA%DB%8C-%D8%A7%D8%B2-joje-twit-04-26)

**Comment by hossein-mohseni** [comment]
سلام
از لحاظ تکنیکال علاوه بر delay بین پکت ها میتونید پروتوکل رندوم برای هر پکت بزارید برای مثال اگر قراره 20 تا پکت ارسال بشه 10 تا tcp و 5 تا udp و مثلا 5 تا هم http ارسال بشه و نوع هر پروتوکل رو میشه از header تشخیص داد.
و یا حتی قبل از اتصال برای هر کلاینت تست delay گرفته بشه و بهترین پروتوکل انتخاب بشه

**Reply by GFW-knocker** [reply]
ما الان به سرور کلودفلر وصلیم و به قوانین  ایپی پایبند باید باشیم
برای سرورهای شخصی حرف شما درسته ولی ایپی رو از بیخ میبندن
کلودفلر رو نمیتونن ببندن چون اگه ببندن 30 درصد نت به چوخ میره

**Comment by hossein-mohseni** [comment]
و به غیر از این روش ها میتونید دیتای اضافه رندومی به پکت ها اضافه کنید تا از DPI جلوگیری بشه (این روش ها برای دور زدن خیلی از فایروال ها در nmap استفاده میشه) و میتونید top 10 روش هایی که برای دور زدن فایروال استفاده میشن رو ترکیب کنید و تست کنید.

به طور خلاصه وقتی روش های موجود الگوریتم مرتب اما پیچیده ای رو استفاده میکنه باز از روی ترتیب میشه اون رو شناسایی کرد ولی وقتی تصادفی باشه خیلی خیلی سختتر میشه.

**Comment by free-the-internet** [comment]
در مورد مشکل یوتیوب ... ورکرنویس ها، شاید بشه با استفاده از ورکر کادفلیر یک پراکسی برای یوتیوب ساخت (شخصی) و هر کسی از آدرس ورکر خودش استفاده کنه برای دیدن یوتیوب. پروژه مشابه: https://github.com/suconghou/u2worker
